UI Clustering: Add management dialog

zyadtaha opened the pull request:
This is the second PR in the GSoC project.

Closes #6698

Changes proposed in this pull request:
- Added a `manage functions` button in the clustering dialog.
![Screenshot from 2024-07-04 16-57-34](https://github.com/OpenRefine/OpenRefine/assets/158375948/baf25571-42d2-4c67-801a-1622adb4103d)
- Clicking the manage functions button opens the management dialog.
![Screenshot from 2024-07-04 16-59-02](https://github.com/OpenRefine/OpenRefine/assets/158375948/ff32d206-7df0-4381-9be9-5266127bd29b)
- Clicking the "add new functions" button opens an expression dialog to enter the function name and its expression.
![Screenshot from 2024-07-04 18-42-47](https://github.com/OpenRefine/OpenRefine/assets/158375948/d5632b06-705d-477e-8b22-61e99905e37f)


Here's a [video demo](https://github.com/OpenRefine/OpenRefine/assets/158375948/b2fe19b0-6aa7-400e-a3d3-b2063211ae7a)

zyadtaha commented:
@wetneb Do you know why the facet tests are failing? They also fail on my master branch, which makes me think it's not related to my PR.

wetneb commented:
That's curious indeed, I don't know what could be causing it. I have opened an issue to investigate it: #6707. We can ignore this failure here.

wetneb commented:
About the PR itself: it looks really great to me! Before diving in nitty-gritty code details, let's see what people think of the design. I have notified people in the old PR and on the forum so that they aware of this exciting new development. :)

tfmorris commented:
I'm not sure what's causing the scatterplot PNG I/O errors, but it looks to me like the actual failing test is:

```
  1) cypress/e2e/project/grid/column/facet/facets.cy.js
       Test opening the clustering from a facet:
     AssertionError: Timed out retrying after 4000ms: expected '<div.dialog-header.ui-draggable-handle>' to contain 'Cluster and edit column "Shrt_Desc"'
      at Context.eval (webpack://openrefine-cypress-test-suite/./cypress/e2e/project/grid/column/facet/facets.cy.js:377:47)
      at runnable.fn (http://127.0.0.1:3333/__cypress/runner/cypress_runner.js:138261:19)
      at callFn (http://127.0.0.1:3333/__cypress/runner/cypress_runner.js:156877:21)
      at Runnable.run (http://127.0.0.1:3333/__cypress/runner/cypress_runner.js:156864:7)
      at <unknown> (http://127.0.0.1:3333/__cypress/runner/cypress_runner.js:162788:30)
      at PassThroughHandlerContext.finallyHandler (http://127.0.0.1:3333/__cypress/runner/cypress_runner.js:4072:23)
      at PassThroughHandlerContext.tryCatcher (http://127.0.0.1:3333/__cypress/runner/cypress_runner.js:1807:23)
      at Promise._settlePromiseFromHandler (http://127.0.0.1:3333/__cypress/runner/cypress_runner.js:1519:31)
```

which sounds like it could possibly be related to this PR since it mentions clustering.

wetneb commented:
Ah, indeed. I just checked and the "Test opening the clustering from a facet" fails on this PR and passes on the master branch on my machine.

ostephens commented:
The design shown here looks good to me. Will add some further comments/questions in the forum thread to avoid clogging up this PR discussion with broader questions

zyadtaha commented:
Here's a [demo](https://github.com/user-attachments/assets/5bc9d30e-013f-4b04-b9e8-f5842198e182) of how the user will add, edit and remove functions so far.
This PR is not finished yet, and I'll keep working on it. If you have any suggestions for improvement, please let me know!

wetneb reviewed (commented):
This is looking really great! Let me know when you need me to review and merge. Based on the comments I understand you want to finish up some things before, no?

zyadtaha commented:
@wetneb I intend to finish this before the review
- [ ] Improve the visuals of the management dialog using the [design system](https://openrefine.org/docs/technical-reference/openRefine-design-system)

Also, I'm thinking of making a separate PR for the "preview clusters" tab to make it easier to review. Do you prefer it included here instead?

zyadtaha commented:
Hi @ostephens,
We have some new designs on the [forum](https://forum.openrefine.org/t/user-defined-clustering-project/1528/27) that I think you might be interested in. I'd love to hear your opinion on them.

wetneb commented:
A separate PR for previewing clusters makes sense.

zyadtaha commented:
**Before the last commit:** 
![](https://github.com/user-attachments/assets/9cd2c6fe-a5dc-4193-b960-a7bf3c170d28)
**After the last commit:** 
![](https://github.com/user-attachments/assets/18a8f7bc-c304-44d0-bb06-c0806fb88b2a)

**Changes made with the last commit:** 
- Added a scroll bar to let users scroll when there are many custom functions.
- Removed the inner line/border between "Name" & "Action"
- Added black borders and a gradient to the "Edit" & "Remove" buttons
- Improved the padding in the dialog
- Fixed column widths so they don’t resize automatically.

Here's a [demo](https://github.com/user-attachments/assets/6a3868f3-327f-4494-8f5f-68ba3b701278) of what it looks like.

zyadtaha commented:
This PR is now ready for review and merge! 
Excited to hear your feedback @wetneb

wetneb commented on main/webapp/modules/core/scripts/dialogs/management-dialog.js:
```diff
@@ -0,0 +1,355 @@
+                }),
+                success: function(data1) {
+                    var _functions = JSON.parse(data1.value);
+                    var langAndExpr = previewWidget.getExpression().split(':');
```
Beware that `:` could be contained in the expression itself, so this is not guaranteed to return two elements.

wetneb reviewed (changes requested):
Very nice! Here are a few comments after trying it out and having a look at the code changes.

I tried adding a keying function, but after doing that it did not show in the list of custom keying functions. In the console I see this error, which is likely related:
```
Uncaught TypeError: _functions is null
    success http://127.0.0.1:3333/project-bundle.js:54145
    fire http://127.0.0.1:3333/project-bundle.js:3225
    fireWith http://127.0.0.1:3333/project-bundle.js:3355
    done http://127.0.0.1:3333/project-bundle.js:9629
    callback http://127.0.0.1:3333/project-bundle.js:9890
    send http://127.0.0.1:3333/project-bundle.js:9909
    ajax http://127.0.0.1:3333/project-bundle.js:9523
    <anonymous> http://127.0.0.1:3333/project-bundle.js:11101
    prop http://127.0.0.1:3333/project-bundle.js:10903
    add http://127.0.0.1:3333/project-bundle.js:54138
    _addFunction http://127.0.0.1:3333/project-bundle.js:54170
    dispatch http://127.0.0.1:3333/project-bundle.js:5147
    handle http://127.0.0.1:3333/project-bundle.js:4951
    add http://127.0.0.1:3333/project-bundle.js:4999
    <anonymous> http://127.0.0.1:3333/project-bundle.js:11525
    prop http://127.0.0.1:3333/project-bundle.js:10903
    on http://127.0.0.1:3333/project-bundle.js:4899
    each http://127.0.0.1:3333/project-bundle.js:385
    each http://127.0.0.1:3333/project-bundle.js:207
    on http://127.0.0.1:3333/project-bundle.js:4898
    on http://127.0.0.1:3333/project-bundle.js:5738
    _addFunction http://127.0.0.1:3333/project-bundle.js:54128
    ManagementDialog http://127.0.0.1:3333/project-bundle.js:53983
    dispatch http://127.0.0.1:3333/project-bundle.js:5147
    handle http://127.0.0.1:3333/project-bundle.js:4951
project-bundle.js:54145:21
```

Perhaps you forgot to handle the case where the preference item does not exist at all in the preference store yet?

I also realized I have some more (minor) comments on the design so I've added them to [the forum thread](https://forum.openrefine.org/t/user-defined-clustering-project/1528/34?u=antonin_d).

zyadtaha commented:
> Perhaps you forgot to handle the case where the preference item does not exist at all in the preference store yet?

![Screenshot from 2024-07-21 22-21-35](https://github.com/user-attachments/assets/0829620b-7768-4896-8f2a-a2c721e8d1f1)
when the data returned is empty, `_functions` is also emtpy and pushes the new function normally. 
I can't reproduce the

zyadtaha commented on main/webapp/modules/core/scripts/dialogs/management-dialog.js:
```diff
@@ -0,0 +1,355 @@
+    
+    $("#management-tabs").tabs({
+        activate: function(event, ui) {
+            if(ui.newTab.text() == "Keying functions"){
```
```
$("#management-tabs").tabs({
        activate: function (event, ui) {
            if (ui.newTab.text() == $.i18n('core-dialogs/keying-functions')) {
                $("#add-new-functions").text($.i18n('core-buttons/add-keying-function'));
            } else {
                $("#add-new-functions").text($.i18n('core-buttons/add-distance-function'));
            }
        }
    });
```
It looks like everything works fine on my end. Could you provide more details about the issue?

zyadtaha commented:
> Perhaps you forgot to handle the case where the preference item does not exist at all in the preference store yet?

Yes, I think that was also the issue causing the dialog to be empty with no placeholder when opened for the first time. Could you check it now and see if the problem is resolved?

wetneb reviewed (commented):
Nice, it works well on my side now!

I have some suggestions of wording changes for the placeholder:
![image](https://github.com/user-attachments/assets/4dd6621f-f0b0-495d-9a94-f9e1a36f6309)
I would say:
* "No functions were found currently" -> "No custom functions" (reason: the "were found currently" feels like it's adding emphasis on the retrieval process out of the backend, but as a user this isn't relevant. Adding the "custom" adjective makes it clear that the emptiness doesn't affect the pre-defined functions)
* "Try adding a new functions below" -> "Add a new function below" (fixes a grammatical error, simplify)

zyadtaha commented:
I have addressed your review comments and added a Cypress test for the Add Custom Function. 
You can review and merge it now. @wetneb

wetneb commented on main/webapp/modules/core/styles/dialogs/clustering-functions-dialog.css:
```diff
@@ -0,0 +1,76 @@
+    color: #333;
+}
+
+.sub-text {
```
Did you intentionally not prefix the two CSS classes above?
Given that they are pretty generic I would say it's not unlikely that they match something else outside of the dialog, for instance introduced by an extension.
The fact that our CSS is split into many different files sadly doesn't mean that we have any isolation.

Instead of changing the class name you could also consider making the CSS selector more specific, such as `#clustering-functions-tabs .main-text`.
